gtsfm/view_graph_estimator/reprojection_error_view_graph_estimator.py

```diff
@@ -0,0 +1,493 @@
+
+        # TODO: cannot loop over triplets to parallelize this, because we do not know its size (single Delayed object).
+        triplet_reports_dict = {}
+        for triplet_idx, triplet in enumerate(triplets_graph):  # sort order guaranteed
+            print(f"On Triplet {triplet_idx}/{len(triplets_graph)}") 
+            triplet_reports_dict[triplet] = self.process_triplet(
+                triplet=triplet,
+                i2Ri1_dict=i2Ri1_dict,
+                i2Ui1_dict=i2Ui1_dict,
+                calibrations=calibrations,
+                corr_idxs_i1i2=corr_idxs_i1i2,
+                keypoints=keypoints,
+                two_view_reports=two_view_reports,
+                cameras_gt=cameras_gt,
+                images=images
+            )
+
+        self.__save_plots(triplet_reports_dict)
+        valid_edges = self.aggregate_valid_edges(triplet_reports_dict)
+        return valid_edges
+
+
+    def estimate_triplets(self, i2Ri1_dict: Dict[Tuple[int, int], Rot3]) -> List[Tuple[int,int,int]]:
+        """ """
+        logger.info("Input number of edges: %d" % len(i2Ri1_dict))
+        input_edges: List[Tuple[int, int]] = i2Ri1_dict.keys()
+        triplets: List[Tuple[int, int, int]] = graph_utils.extract_cyclic_triplets_from_edges(input_edges)
+
+        N = len(triplets)
+        logger.info("Number of triplets: %d" % N)
+
+        return triplets
+
+
+    def aggregate_valid_edges(self, triplet_reports_dict: Dict[Tuple[int,int,int], TripletOptimizationReport]):
+        """
+        """
+        valid_edges = set()
+        for (i0,i1,i2), report in triplet_reports_dict.items():
+            if not report.is_success:
+                continue
+            valid_edges.add((i0, i1))
+            valid_edges.add((i1, i2))
+            valid_edges.add((i0, i2))
+
+        return valid_edges
+
+
+    def process_triplet(
+        self,
+        triplet: Tuple[int,int,int],
+        i2Ri1_dict: Dict[Tuple[int, int], Rot3],
+        i2Ui1_dict: Dict[Tuple[int, int], Unit3],
+        calibrations: List[Cal3Bundler],
+        corr_idxs_i1i2: Dict[Tuple[int, int], np.ndarray],
+        keypoints: List[Keypoints],
+        two_view_reports: Dict[Tuple[int, int], TwoViewEstimationReport],
+        cameras_gt: Optional[List[PinholeCameraCal3Bundler]] = None,
+        images = None,
+    ) -> TripletOptimizationReport:
+        """ """
+        if triplet is None:
+            return TripletOptimizationReport(triplet=triplet)
+        i0, i1, i2 = triplet
+
+        # immediately reject bad cycles
+        cycle_error = comp_utils.compute_cyclic_rotation_error(
+            i1Ri0=i2Ri1_dict[(i0, i1)], i2Ri1=i2Ri1_dict[(i1, i2)], i2Ri0=i2Ri1_dict[(i0, i2)]
+        )
+        if cycle_error > MAX_ALLOWED_ROT_CYCLE_ERROR_DEG:
+            logger.info(f"Immediately rejected for cycle error {cycle_error:.1f} deg.")
+            return TripletOptimizationReport(triplet=triplet)
```

Comment:
I feel we should have a cascade of view graph estimators. Will lead to minimal logic in this file.